### PR TITLE
Audit use of saveLayer

### PR DESCRIPTION
**Pull request by Hixie**
It has recently come to light that we may have been ordering saveLayer
calls and clip calls incorrectly.

### CODE REVIEW

**Comment by Hixie** [comment]
cc @chinmaygarde @abarth

File: packages/flutter/lib/src/material/time_picker.dart [hunk 1]
```diff
@@ -429,7 +429,7 @@ class _DialPainter extends CustomPainter {
       center: focusedPoint, radius: focusedRadius
     );
     canvas
-      ..saveLayer(focusedRect, new Paint())
+      ..save()
```
Comment (abarth) [inline packages/flutter/lib/src/material/time_picker.dart]: Did you check that this works properly?

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -355,10 +355,12 @@ class PaintingContext {
     if (needsCompositing) {
       pushLayer(new ClipRRectLayer(clipRRect: offsetClipRRect), painter, offset, childPaintBounds: offsetBounds);
     } else {
-      canvas.saveLayer(offsetBounds, _defaultPaint);
+      canvas.save();
       canvas.clipRRect(offsetClipRRect);
+      canvas.saveLayer(offsetBounds, _defaultPaint);
```
Comment (abarth) [inline packages/flutter/lib/src/rendering/object.dart]: Consider `..` operator.

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/material/time_picker.dart (hunk 1 above)
Comment (Hixie) [inline packages/flutter/lib/src/material/time_picker.dart]: Assuming I was looking at the right part of the dial, it didn't seem to make any difference.

**Comment by abarth** [comment]
Why are the extra save/restores needed?  I thought from your examples that the clip was applied to the next operation and then popped.

**Comment by abarth** [comment]
Nevermind.  I re-read the examples.

File: packages/flutter/lib/src/rendering/object.dart (hunk 2 above)
Comment (Hixie) [inline packages/flutter/lib/src/rendering/object.dart]: applied liberally

**Comment by Hixie** [comment]
(to answer the question for anyone else who wonders: the save before the clip is needed so that the clip can be restored, the saveLayer after the clip (if any) is needed to cause the graphics to composite correctly, and needs its own restore to trigger the composition.)

**Comment by Hixie** [comment]
will land on green

**Pull request merged**